### DISCUSSION TITLE
Automatic startup programs all open in same workspace.

### DISCUSSION THREAD

**Discussion by satisfiedghost** [i3/i3#5790] [Q&A] [answered]
I'm trying to make it so a default set of programs (browser, media player, message apps) all open in the layout I manually configure every time I boot.

I have something much like the following in my `~/.config/i3/config`:

```
exec --no-startup-id i3-msg 'workspace 1; exec firefox'
exec --no-startup-id i3-msg 'workspace 2; exec joplin'
exec --no-startup-id i3-msg 'workspace 3; exec discord'
exec --no-startup-id i3-msg 'workspace 3; exec telegram'
exec --no-startup-id i3-msg 'workspace 3; exec signal-desktop'
exec --no-startup-id i3-msg 'workspace 3; layout stacking'
exec --no-startup-id i3-msg 'workspace 7; exec spotify'
exec --no-startup-id i3-msg 'workspace 1'
```

All the programs do start, but only in workspace 1 with the default vertical tiling.

Per some suggestions I bumped into I've tried appending a `; workspace 1` to the command in each line but the behavior is unchanged.

NB: I've noticed if I open an application and then switch to another workspace before its window appears, it will load on whatever the active workspace is. Perhaps that's happening here, where these commands are run near instantaneously and by the time they actually get loaded I'm back to workspace 1, and all their windows get loaded there?

**Comment by okraits** [comment, marked as answer]
Please use the _assign_ directive https://i3wm.org/docs/userguide.html#assign_workspace or layout saving https://i3wm.org/docs/layout-saving.html to get deterministic results.

**Reply by satisfiedghost** [reply]
Brilliant, the layout saving worked a treat for me. Thanks!

Note to anyone who might find this and run into headaches with Discord specifically: 

Discord starts with a floating updater window that has the same `class`, `instance`, and `window_role` fields at the actual application. This will get swallowed by the empty window created by the `append_layout` command in your `config`. This window then closes and the full application opens up, but will open in your current workspace as the Discord updater has closed the window that swallowed it.

I was able to get around this by using the title `"title": "^Friends\\ \\-\\ Discord$"` in the `swallows` field of the json produced by `i3-save-tree` which feels like a hack as it's relying on the default behavior being that Discord opens to the friends section, but it was the only unique property i3 could use to differentiate them.
